src/main/haskell/matching-logic/app/Main.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Main where
+import           Control.Applicative                           (some)
+import           Control.Monad.IO.Class                        (liftIO)
+import           Control.Monad.State.Strict                    (MonadState (..),
+                                                                StateT,
+                                                                runStateT)
+import           Data.Char                                     (isAlphaNum,
+                                                                isSpace)
+import           Data.List                                     (isPrefixOf,
+                                                                isSuffixOf)
+import qualified Data.Map.Strict                               as Map
+import qualified Data.Set                                      as Set
+import           Data.Text
+
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+
+import           Data.Text.Prettyprint.Doc                     hiding (space)
+import qualified Data.Text.Prettyprint.Doc                     as Doc
+
+import           Data.Reflection
+
+import           System.Console.Haskeline
+
+import           Kore.MatchingLogic.AST
+import           Kore.MatchingLogic.AST.Syntax
+import           Kore.MatchingLogic.HilbertProof
+import           Kore.MatchingLogic.ProofSystem.Minimal
+import           Kore.MatchingLogic.ProofSystem.Minimal.Syntax (parseId,
+                                                                parseMLRuleSig)
+import           Kore.MatchingLogic.ProverRepl
+import           Kore.MatchingLogic.Signature.Simple
+
+-- Todo: Parsing Formula as Text. Hook to Kore Parser
+parseName :: Parser Text
+parseName = takeWhile1P Nothing isAlphaNum <* space
+
+pCommand :: (Reifies s ValidatedSignature)
+         => Parser (Command Text
+                    (MLRuleSig (SimpleSignature s) Text)
+                    (WFPattern (SimpleSignature s) Text))
+pCommand = parseCommand parseName parseFormula parseRule
+  where
+    parseFormula = simpleSigPattern parseName parseName parseName
+    parseLabel = simpleSigLabel parseName
+    parseSort = simpleSigSort parseName
+    parseRule = parseMLRuleSig parseSort parseLabel parseName parseName
+
+
+proveCommand :: (Reifies sig ValidatedSignature)
+             => proxy (SimpleSignature sig)
+             -> IO (ProverState Text (MLRuleSig (SimpleSignature sig) Text) (WFPattern (SimpleSignature sig) Text))
```

Comment:
Several things you are complaining about are not part of our style guide. Let's not hold this merge up further on style suggestions that are not even part of our style guide.

* 80 is recommended, but one of the adopted changes from the Serokell/Cardano guide is raising the hard limit to 100.
* The correct indent in the type signature is not 4, that's for "code blocks", and this exact alignment of type signatures (and constructor declerations) is shown in example code in all the linked style guides that ours incorporates by reference, including Tibbel's original.
* Putting a newline after `=` is required nowhere, and several examples in the style guide show the opposite (e.g, under "Hanging Labdas").
* Using full names rather than abbreviations for local names is nowhere suggested. I'll try not to abbreviate too much.

* `WFPattern` should indeed be documented (and stands for well-formed pattern), but I don't expect to fit much more then mnemonics in an identifier without becoming excessively large.
* Restricting letter-named arguments is actually a documented requirement and `autoSort` is a bit tricky, I'll improve that.
* Documentation is good and I'll see what I can do by tomorrow, but I'm willing to take on more debt here to get this on master.

* Naming constructor arguments is a fair suggestion, but I don't see `<mv>` in your example code.

Reply:
What do you think our style guide is? I'm asking because our style guide does not yet have a written description. There are several things that are related to our style guide:
* The style guide that is used as a base by IOHK, and which influenced ours.
* The style used in the current code (which is mostly, but not fully, consistent).
* The style that our modified version of hlint uses (I'm not fully happy with it, but we could adopt that).

Now, I think that _all_ of the things above are part of the style that was used until now, and that we made an explicit effort to follow them (although the actual style developed with time and earlier code may not always follow the later style). I'm not sure which of my suggestions you think were not part of the style of the current code, so I can't provide more details.

80/100 columns: I don't mind that much having an 100 chars limit, if we decide that we actually want that, but our current style uses 80 characters. With 80 characters one can fit three source code windows (or two code windows and one console) side-by-side on many modern monitors, and, given the poor IDE support for Haskell compared with languages like Java, I actually found this to be useful in many cases. If I'm the only one that thinks 3 side by side windows are useful, let's raise the limit to 100.

Indentation: The correct indent is what we think it should be. There are two problems with putting the "::" after the function name and aligning all the arrows with it. First, it uses a lot of the available horizontal space (be it 80 or 100 columns), and I don't see what benefit we would get out of it. Second, given the poor Haskell IDE support, having indent levels which are multiples of
(say) 4 helps a lot when aligning the code. And it does not matter if some IDEs/editors have no problem with this, because, most likely, they fail/are suboptimal in other (important) ways so you can't simply ask everyone to switch to your preferred IDE/editor.

Also, the `proveCommand` function has a really short name, but longer function names would make things much worse. Anyway, assuming that there is some reason for aligning things at the end of the function name, we could change our current style guide, but then we should also change the existing code to conform to the new style guide. Would you mind doing that?

Newline after "=": Again, you're expecting a written style guide, when none exists. Also, the only example for our (unwritten) style guide is the current code, so I don't think you'll find many examples showing the opposite.

Using full names rather than abbreviations: this was one of the places where I used "prefer", because I wanted to make a weaker request than in other places (I'm not sure that I used the proper word, but that was my intent).The current code sort of follows this rule. The main point here was to decrease the mental effort needed to read the code.

"[...] fit much more then mnemonics in an identifier without becoming excessively large" - The same point as above - if the abbreviations (I don't think mnemonic is the right word here, but my English is not that good) are well known (e.g. km or kWh) then there's no point in writing anything longer, but for mnemonics/abbreviations which are well known only to you or only to a small group, but not to most people involved in this project, you're making things harder to read. Documentation makes the problem smaller, but it does not solve it fully, since it requires moving to a different file in order to understand a name at its usage place. The full problem of when to use abbreviations and when not to is more complex, so I won't attempt it, I'll just be content with whatever you choose: documentation or longer name.

Documentation debt: That's fine in principle, but it will take me much longer to review your code, so, although it will probably still be faster than to write documentation, I'm not sure you'll gain much. Also consider that Traian will also use more time to review this, and there is a good chance that my extra time + Traian's extra time are longer than the time needed to write documentation.

I can't speak for Traian, but I'll continue reviewing this and I'll be waiting a pull request with documentation after this one is merged.

`<mv>` - this is probably one of those cases when I forgot do backtick-escape something and github ate my message (or maybe it was a copy-paste error). The other form for data declarations was:
`data C = C1 <mv> | C2 <mv> | ... | Cn <mv>`
where `n>=1` and `<mv>` is a non-terminal, i.e. in the same declaration some of them can be empty and some can be single types.

Reply:
Our style guide is posted here: https://github.com/kframework/kore/wiki/Coding-Standards
We've adopted code formatting guidelines wholesale from the linked sources, and added a few points about testing and build infrastructure. I'm assuming any of the examples in those guides show acceptable code. Overall, I think we mostly want a consistent style, with a bit of alignment for readability. If we can set up hindent or something like that to impose an acceptable style that's fine with me.

I've set up my editor to mark 80 columns now. On my monitors two 90-column files barely fit side by side, but that's a local thing (and I have multiple monitors to work with).

Restricting data declarations like that is a bad idea. There is rarely much to gain from splitting sum and product types like that.

Reply:
I thought I mentioned it to people when it was written after the discussions on Slack, but I don't see where.

For type declarations I think using record syntax in declarations is fine to help disambiguate, but it's rarely work splitting out a separate type and duplicating instance declarations and so on. You easily end up with 40 or 60 characters of your 80 column line used up just by identifier names when you add too many layers. Also, we should use `DisambiguateRecordFields` - it allows names to be reused between records, with no need for disambiguation when they are being used within pattern matching or with the record constructor.